src/posts/2022-11-03-constellation/index.md

```diff
@@ -0,0 +1,148 @@
+ 
+The Cilium community was very kind and helpful with our problem. They were quick to respond and helped us with the implementation of the strict mode. In the discussions with the Cilium maintainers, they suggested a location [in the datapath](https://docs.cilium.io/en/v1.12/concepts/ebpf/lifeofapacket/#life-of-a-packet) where we should add our [IP-based filter](https://github.com/edgelesssys/cilium/blob/d465085df155502267589f782801c63f3d08dd40/bpf/lib/strict.h). After some trial and error, we finally found the correct spots in `bpf/bpf_host.c` and [`bpf/lib/encap.h`](https://github.com/edgelesssys/cilium/blob/feat/v1.12/ciliumStrictMode/bpf/lib/encap.h#L235), which made our code pass the test. For the eBPF connoisseurs, we’ve listed the actual packet filter code below. If you’re unsure what you’re seeing there – it filters packets based on IPv4 subnets. What? We never said it’s rocket science.
+ 
+```c
+// strict_allow checks whether the packet is allowed to pass through the strict mode.
+static __always_inline bool
+strict_allow(struct __ctx_buff *ctx)
+{
+ void *data, *data_end;
+ __u16 proto = 0;
+ bool __maybe_unused in_strict_cidr = false;
+#ifdef ENABLE_IPV4
+ struct iphdr *ip4;
+#endif
+    if (!validate_ethertype(ctx, &proto))
+     return true;
+
+ switch (proto)
+ {
+#ifdef ENABLE_IPV4
+ case bpf_htons(ETH_P_IP):
+     if (!revalidate_data(ctx, &data, &data_end, &ip4))
+         return true;
+
+     
+     switch (ip4->protocol)
+     {
+     case IPPROTO_TCP:
+     case IPPROTO_UDP:
+         in_strict_cidr = ipv4_is_in_subnet(ip4->daddr, STRICT_IPV4_NET, STRICT_IPV4_NET_SIZE);
+         in_strict_cidr &= ipv4_is_in_subnet(ip4->saddr, STRICT_IPV4_NET, STRICT_IPV4_NET_SIZE);
+         return !in_strict_cidr;
+     case IPPROTO_ICMP:
+         return true;
+     case IPPROTO_ICMPV6:
+         return true;
+     default:
+         return false;
+     }
+
+     break;
+#endif
+ default:
+     return true;
+ }
+}
+```
+Description: Basic eBPF IPv4 packet filter based on CIDR.
+ 
+The full implementation can currently be found in [our Cilium fork](https://github.com/edgelesssys/cilium/tree/feat/v1.12/ciliumStrictMode), and is used and configured automatically by Constellation. While our current implementation is somewhat specific to our use case, we’re confident that it can help protect the workload traffic of Cilium users on a wider basis.
+ 
+The reason why we did not upstream our change yet, is that soon the datapath will be changed when Cilium rolls out [Node-to-Node](https://github.com/cilium/cilium/pull/19401) encryption in their [next release](https://github.com/cilium/cilium/milestone/37). Then we’ll re-evaluate if the strict mode is still needed and how our implementation needs to be adapted to be beneficial for all Cilium users.
```

Comment:
Node-to-node encryption for IPsec isn't on the roadmap yet. In addition, even once node-to-node encryption is implemented, I think it will still be useful to have strict mode. Not everyone will enable node-to-node encryption because that will lead to more overhead.

We would very much welcome a submission of this extension upstream! I think it would also make for a stronger blog post if it was upstreamed.

Reply:
Thanks, @pchaigno. We'll prepare it for upsteaming and open a PR.
I'll adjust the blog post accordingly.
CC @3u13r